### PR TITLE
[AutoDiff] Support differentiation of wrapped properties.

### PR DESCRIPTION
**Pull request by dan-zheng**
Differentiation now correctly supports stored properties with attached property wrappers.

### `Differentiable` derived conformances

`Differentiable` conformance derivation now "peers through" property wrappers.
This behavior is always desirable.

Synthesized `TangentVector` structs contain wrapped properties' `TangentVector`s
as stored properties, not wrappers' `TangentVector`s.

It is irrelevant whether property wrapper types conform to `Differentiable`.
Property wrapper types are required to provide a setter for `var wrappedValue`,
which is needed to synthesize `mutating func move(along:)`.

### Differentiation transform

The differentiation transform is also updated to support differentiation of
wrapped value accessors.

This is done via a new pullback generation code path for "semantic member
accessors". "Semantic member accessors" are accessors attached to member
properties that have a corresponding tangent stored property in the parent
`TangentVector` type.

These accessors have special-case pullback generation based on their semantic
behavior. Currently, only getters and setters are supported.

This special-case pullback generation is currently used for stored property
accessors and property wrapper wrapped value accessors. In the future, it can
also be used to support `@differentiable(useInTangentVector)` computed
properties: SR-12636.

User-defined accesors cannot use this code path because they may use custom
logic that does not semantically perform a member access.

### Example

```swift
import _Differentiation

@propertyWrapper
struct Wrapper<Value> {
  var wrappedValue: Value
}

struct Struct {
  @Wrapper var x: Float = 0
  // Compiler generates:
  // var _x: Wrapper<Float>
  // var x: Float {
  //   get { _x.wrappedValue }
  //   set { _x.wrappedValue = newValue }
  // }
}

extension Struct: Differentiable {
  // Compiler now synthesizes:
  // struct TangentVector: Differentiable & AdditiveArithmetic {
  //   var x: Float
  //   ...
  // }
}

@differentiable
func testWrappedValueAccessors(_ s: Struct, _ x: Float) -> Float {
  var s = s
  // `Struct.x.get` and `Struct.x.set` are both differentiable.
  s.x = s.x * x
  // `modify` accessors are not yet supported, so the line below triggers a diagnostic.
  // s.x *= x
  return s.x
}
```

---

Resolves all of SR-12637 that can be supported today.

### CODE REVIEW

File: lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp [hunk 1]
```diff
@@ -510,6 +512,232 @@ void PullbackEmitter::addToAdjointBuffer(SILBasicBlock *origBB,
...
+    return false;
+  auto *accessor = dyn_cast<AccessorDecl>(dc->getAsDecl());
+  if (!accessor)
+    return false;
+  // Currently, only getters and setters are supported.
+  // TODO(SR-12640): Support `modify` accessors.
+  if (accessor->getAccessorKind() != AccessorKind::Get &&
+      accessor->getAccessorKind() != AccessorKind::Set)
+    return false;
+  // Currently, only implicit accessors are supported.
```
Comment (rxwei) [inline lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp:535]: implicit accessors include getters/setters synthesized for stored properties. I'm not sure checking for implicitness is enough. I think we should check whether `AccessorDecl`'s parent decl is a `VarDecl` that has a property wrapper on it.

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -2723,13 +2723,22 @@ WARNING(differentiable_nondiff_type_implicit_noderivative_fixit,none,
...
-      (Identifier, Type, Identifier, bool))
-WARNING(differentiable_let_property_implicit_noderivative_fixit,none,
+      (/*propName*/ Identifier, /*propType*/ Type, /*nominalName*/ Identifier,
+       /*nominalCanDeriveAdditiveArithmetic*/ bool))
+WARNING(differentiable_immutable_wrapper_implicit_noderivative_fixit,none,
       "synthesis of the 'Differentiable.move(along:)' requirement for %1 "
-      "requires all stored properties to be mutable; use 'var' instead, or add "
-      "an explicit '@noDerivative' attribute"
+      "requires all stored properties not marked with `@noDerivative` to be "
+      "mutable; add '%0.wrappedValue.set', or add an explicit '@noDerivative' "
```
Comment (rxwei) [inline include/swift/AST/DiagnosticsSema.def:2731]: * `.set` is not a user-facing identifier today. I'd suggest describing this in plain English, e.g. "a setter for 'Foo.wrappedValue'".
* Suggesting that users add a setter isn't always helpful. Users are most likely using a property wrapper that they did not define, and adding a setter isn't something that can be done retroactively. I'd suggest removing this suggestion.

File: lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp]: I made the check more explicit in https://github.com/apple/swift/pull/31173/commits/49c4441fc48762261467e7d3893fb6942c051ab1.

I intentionally included stored property accessors to validate the new code path, and because the new code path has more room for micro-optimization. But I'm fine with reverting it.

Comment (dan-zheng) [reply lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp]: Another idea: we can lift the following check into `isSemanticMemberAccessor`: "`TangentVector` must be a struct that contains a stored property with the same name and corresponding `TangentVector` type of the given property".

If we do this, the new code path will be triggered only if it is guaranteed not to emit a non-differentiability error, which may slightly improve diagnostic locations. I'd prefer to investigate this in a follow-up.

File: include/swift/AST/DiagnosticsSema.def (hunk 2 above)
Comment (dan-zheng) [reply include/swift/AST/DiagnosticsSema.def]: > * I'd suggest removing this suggestion.

Done in https://github.com/apple/swift/pull/31173/commits/f4b389c30b8726977b1ba325cdded9bf0a1c2572.

Alternatively, we could conditionally suggest "adding a setter for 'Wrapper.wrappedValue'" iff the wrapper type is defined in the same file or module.

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
Force-pushed to squash minor commits. Testing now.

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
There's an unrelated macOS validation test failure, for other PRs too.
```
00:52:56 Failing Tests (1):
00:52:56     Swift(macosx-x86_64) :: api-digester/compare-dump-abi.swift
```

I'll just trigger smoke tests.
@swift-ci Please smoke test macOS

File: lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp [hunk 3]
```diff
@@ -510,6 +512,242 @@ void PullbackEmitter::addToAdjointBuffer(SILBasicBlock *origBB,
...
+
+/// Returns true if the given original function is a "semantic member accessor".
+static bool isSemanticMemberAccessor(SILFunction *original) {
+  auto *dc = original->getDeclContext();
+  if (!dc)
+    return false;
+  auto *decl = dc->getAsDecl();
+  if (!decl)
+    return false;
+  auto *accessor = dyn_cast<AccessorDecl>(dc->getAsDecl());
```
Comment (theblixguy) [inline lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp:527]:
```suggestion
  auto *accessor = dyn_cast<AccessorDecl>(decl);
```

File: lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp [hunk 4]
```diff
@@ -510,6 +512,242 @@ void PullbackEmitter::addToAdjointBuffer(SILBasicBlock *origBB,
...
+    return true;
+  // Return true for properties that have attached property wrappers.
+  if (varDecl->hasAttachedPropertyWrapper())
+    return true;
+  // Otherwise, return false.
+  // User-defined accessors can never be supported because they may use custom
+  // logic that does not semantically perform a member access.
+  // TODO(SR-12636): Support `@differentiable(useInTangentVector)` computed
+  // properties.
+  return false;
```
Comment (theblixguy) [inline lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp:550]: You could just `return varDecl->hasAttachedPropertyWrapper();` here.

File: lib/Sema/DerivedConformanceDifferentiable.cpp [hunk 5]
```diff
@@ -39,14 +40,23 @@ getStoredPropertiesForDifferentiation(NominalTypeDecl *nominal, DeclContext *DC,
...
+    // Peer through property wrappers: use original wrapped properties instead.
+    if (auto *originalProperty = vd->getOriginalWrappedProperty()) {
+      // Skip property wrappers that do not define `wrappedValue.set`.
+      // `mutating func move(along:)` cannot be synthesized to update these
+      // properties.
+      auto *wrapperDecl =
+          vd->getInterfaceType()->getNominalOrBoundGenericNominal();
+      auto *wrappedValueDecl =
+          wrapperDecl->getPropertyWrapperTypeInfo().valueVar;
+      if (!wrappedValueDecl->getAccessor(AccessorKind::Set))
```
Comment (theblixguy) [inline lib/Sema/DerivedConformanceDifferentiable.cpp:52]: I think you can perform this check on `originalProperty` directly - it won’t have a setter if the `wrappedValue` doesn’t have one.

File: lib/Sema/DerivedConformanceDifferentiable.cpp [hunk 6]
```diff
@@ -484,20 +494,52 @@ static void addAssociatedTypeAliasDecl(Identifier name, DeclContext *sourceDC,
...
+      if (!wrappedValueDecl->getAccessor(AccessorKind::Set)) {
+        auto loc =
+            originalProperty->getAttributeInsertionLoc(/*forModifier*/ false);
+        Context.Diags
+            .diagnose(
+                loc,
+                diag::
+                    differentiable_immutable_wrapper_implicit_noderivative_fixit,
+                wrapperDecl->getNameStr(), nominal->getName(),
+                nominalCanDeriveAdditiveArithmetic)
```
Comment (theblixguy) [inline lib/Sema/DerivedConformanceDifferentiable.cpp:528]: The formatting here looks a little strange to me.

File: lib/Sema/DerivedConformanceDifferentiable.cpp (hunk 5 above)
Comment (dan-zheng) [reply lib/Sema/DerivedConformanceDifferentiable.cpp]: Done. Thanks for the review!

File: lib/Sema/DerivedConformanceDifferentiable.cpp (hunk 6 above)
Comment (dan-zheng) [reply lib/Sema/DerivedConformanceDifferentiable.cpp:525]: I just use `git clang-format` as an auto-formatter.

File: lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp]: Done, thanks.

**Comment by dan-zheng** [comment]
I pushed some final fixes to finish wrapped property differentiation:

> [AutoDiff] Finish wrapped property differentiation support.
> 
> Add special-case VJP generation support for "semantic member accessors".
> This is necessary to avoid activity analysis related diagnostics and simplifies
> generated code.
> 
> Fix "wrapped property mutability" check in `Differentiable` derived conformnances.
> This resolves SR-12642.
> 
> Add e2e test using real world property wrappers.

This PR is ready to land now.

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Review by rxwei — APPROVED** [review]
